### DISCUSSION TITLE
How to install python package directly from GitHub?

### DISCUSSION THREAD

**Discussion by asmith26** [prefix-dev/pixi#1710] [Q&A] [answered]
What's the equivalent to `pip install git+https://github.com/huggingface/parler-tts.git`?

**Comment by asmith26** [comment, marked as answer]
This seems to work/be the correct format: `pixi add --pypi "parler-tts @ git+https://github.com/huggingface/parler-tts.git"`

**Reply by Shobhit0109** [reply]
How can we install globally from github?

**Reply by matthewfeickert** [reply]
@Shobhit0109 at the moment `pixi global` doesn't have `pipx` like functionality and it isn't on the roadmap given https://github.com/prefix-dev/pixi/issues/565#issuecomment-2286273461. So to achieve what you want you'd do

```console
# pixi global install git  # if you somehow don't have Git
$ pixi global install pipx
$ pipx install 'git+https://github.com/huggingface/parler-tts.git'
```

However, if you do that you'll get the notice

```
No apps associated with package parler_tts. Try again with '--include-deps' to include apps of dependent packages, which are listed above. If you are attempting to install a library, pipx should not be used.
Consider using pip or a similar tool instead.
```

so apparently `parler-tts` doesn't provide any CLI API. Given this, I'm not sure what your use case is of wanting to try to install it globally.

**Reply by lucascolley** [reply]
see https://github.com/prefix-dev/pixi/issues/3725 for global specific case

**Comment by PiotrCzapla** [comment]
Hi guys this doesn't seems to work anymore:
```
pixi add --pypi "parler-tts @ git+https://github.com/huggingface/parler-tts.git"
  × Expected package name starting with an alphanumeric character, found `@`
  │ @
  │ ^
  ```

**Reply by matthewfeickert** [reply]
@PiotrCzapla That works for me on 

```console
$ pixi --version
pixi 0.33.0
```

```console
$ pixi init
✔ Created /tmp/pixi-debug/pixi.toml
$ pixi add 'python<3.13'
✔ Added python<3.13
$ pixi add --pypi "parler-tts @ git+https://github.com/huggingface/parler-tts.git"
  × failed to solve the pypi requirements of 'default' 'linux-64'
  ╰─▶ Build backend failed to determine metadata through `prepare_metadata_for_build_wheel` with exit status: 1
      --- stdout:
      
      --- stderr:
      Traceback (most recent call last):
        File "<string>", line 14, in <module>
        File "/home/feickert/.cache/rattler/cache/uv-cache/builds-v0/.tmpKO6HKs/lib/python3.12/site-packages/setuptools/build_meta.py", line 373, in prepare_metadata_for_build_wheel
          self.run_setup()
        File "/home/feickert/.cache/rattler/cache/uv-cache/builds-v0/.tmpKO6HKs/lib/python3.12/site-packages/setuptools/build_meta.py", line 516, in run_setup
          super().run_setup(setup_script=setup_script)
        File "/home/feickert/.cache/rattler/cache/uv-cache/builds-v0/.tmpKO6HKs/lib/python3.12/site-packages/setuptools/build_meta.py", line 318, in run_setup
          exec(code, locals())
        File "<string>", line 354, in <module>
        File "<string>", line 87, in get_ext_modules
      ModuleNotFoundError: No module named 'numpy.distutils'
      ---

```

(where this is obviously failing as `numpy.distutils` has been deprecated for some time, but that's not a `pixi` issue.)

**Reply by fnands** [reply]
In case someone runs into this later, I had the same issue as @PiotrCzapla . 

With `pixi 0.40.3` the solution was to remove the spaces around the `@`, i.e.: 

```
pixi add --pypi "parler-tts@git+https://github.com/huggingface/parler-tts.git"
```

**Comment by SheldonWBM** [comment]
I am experimenting with Pixi and seeing which projects a Pixi package manager can be applied to. 
```
pixi add --PyPI "label-studio-ml @ git+https://github.com/HumanSignal/label-studio-ml-backend.git"
  x failed to solve the pypi requirements of 'default' 'linux-64'
  |-> Failed to download and build: `label-studio-sdk @ git+https://github.com/HumanSignal/label-studio-sdk.git`
  |-> Failed to extract static metadata from `pyproject.toml`
  |-> `pyproject.toml` is using the `[project]` table, but the required `project.name` field is not set.
  `-> TOML parse error at line 30, column 2
         |
      30 | [project.urls]
         |  ^^^^^^^
      missing field `name`
```
When I move the dependency from `[pypi-dependencies]` to `[dependencies]` I have this error
```
label-studio-ml = { git = "https://github.com/HumanSignal/label-studio-ml-backend.git" }
    ---- source dependencies are not allowed yet
```
`label-studio-ml` is out-of-date on PyPi and does not exist in Conda. So, for this project. Being able to install from GitHub or a zip file is essential.

I am checking out Pixi (for fun), as a guest speaker mentioned it in a Posit Data Science Hangout.

**Reply by matthewfeickert** [reply]
@SheldonWBM The issue here isn't `pixi` but that the packages you're trying to install are not in good shape. You can see with the following example that things work fine for well maintained packages like `hist`:

```
mkdir example && cd example
pixi init
pixi add python
pixi add --pypi "hist @ git+https://github.com/scikit-hep/hist.git"
```

will result in a `pixi.toml` like

```toml
[project]
authors = ["Matthew Feickert"]
channels = ["conda-forge"]
description = "Add a short description here"
name = "example"
platforms = ["linux-64"]
version = "0.1.0"

[tasks]

[dependencies]
python = ">=3.13.0,<3.14"

[pypi-dependencies]
hist = { git = "https://github.com/scikit-hep/hist.git" }
```

and a valid `pixi.lock.`

As your original error message shows though

```
pixi add --PyPI "label-studio-ml @ git+https://github.com/HumanSignal/label-studio-ml-backend.git"
  x failed to solve the pypi requirements of 'default' 'linux-64'
  |-> Failed to download and build: `label-studio-sdk @ git+https://github.com/HumanSignal/label-studio-sdk.git`
  |-> Failed to extract static metadata from `pyproject.toml`
  |-> `pyproject.toml` is using the `[project]` table, but the required `project.name` field is not set.
  `-> TOML parse error at line 30, column 2
         |
      30 | [project.urls]
         |  ^^^^^^^
      missing field `name`
```

[`HumanSignal/label-studio-ml-backend` has `label-studio-sdk @ git+https://github.com/HumanSignal/label-studio-sdk.git` as an install requirement](https://github.com/HumanSignal/label-studio-ml-backend/blob/31bb893abde3372a82a9907d771c102fc9065b67/requirements.txt#L7) and [HumanSignal/label-studio-sdk doesn't have valid `pyproject.toml` project metadata](https://github.com/HumanSignal/label-studio-sdk/blob/505bf58261478e9b0babb53a4815ccf8c4bd2ed9/pyproject.toml), but instead only has `poetry` metadata. As `pixi` is using `uv` under the hood to do the install, which is going to be strict about metadata and not guess, this is what is bringing up the error.

```console
$ docker run --rm -ti python:3.12 bash
root@96ccb3619cc7:/# python -m venv venv && . venv/bin/activate && python -m pip --quiet install uv

[notice] A new release of pip is available: 24.2 -> 24.3.1
[notice] To update, run: pip install --upgrade pip
(venv) root@96ccb3619cc7:/# uv pip install 'label-studio-sdk @ git+https://github.com/HumanSignal/label-studio-sdk.git'
Using Python 3.12.7 environment at /venv
 Updated https://github.com/HumanSignal/label-studio-sdk.git (505bf58)
  × Failed to download and build `label-studio-sdk @ git+https://github.com/HumanSignal/label-studio-sdk.git`
  ├─▶ Failed to extract static metadata from `pyproject.toml`
  ├─▶ `pyproject.toml` is using the `[project]` table, but the required `project.name` field is not set.
  ╰─▶ TOML parse error at line 30, column 2
         |
      30 | [project.urls]
         |  ^^^^^^^
      missing field `name`

(venv) root@96ccb3619cc7:/#
```

So only `poetry` and `pip` can install that package as it currently is. This is a problem of the tools you want to install, not the installers. I suggest you open up an Issue with https://github.com/HumanSignal/label-studio-sdk to let them know they're limiting who can install them given invalid metadata and an Issue with https://github.com/HumanSignal/label-studio-ml-backend asking if they will create an updated release to PyPI.

---

For clarity, I'm just a person on the internet. I'm not part of the `pixi` team.

**Reply by SheldonWBM** [reply]
`hist`, `numpy` have a `[project]` field in the pyproject.toml for example:
```
[project]
name = "numpy"
version = "2.2.0.dev0"
# TODO: add `license-files` once PEP 639 is accepted (see meson-python#88)
license = {file = "LICENSE.txt"}

description = "Fundamental package for array computing in Python"
authors = [{name = "Travis E. Oliphant et al."}]
maintainers = [
    {name = "NumPy Developers", email="numpy-discussion@python.org"},
]
requires-python = ">=3.10"
```
The `label-studio-sdk` repository does not yet have a `[project]` field, it has a `[tool.poetry]` field with similar info.

**Reply by matthewfeickert** [reply]
> The `label-studio-sdk` repository does not yet have a `[project]` field, it has a `[tool.poetry]` field with similar info.

@SheldonWBM not quite. It uses [`[project.urls]`](https://github.com/HumanSignal/label-studio-sdk/blob/505bf58261478e9b0babb53a4815ccf8c4bd2ed9/pyproject.toml#L30-L31), as the error message shows, which is invalid without the rest of the project table. Poetry doesn't adhere to the standards and does its own thing, which causes problems like this.

**Reply by SheldonWBM** [reply]
Update: I was able to work around it.
```
[tasks]
install-label-studio-ml = "pip install git+https://github.com/HumanSignal/label-studio-ml-backend.git"
install-label-studio-sdk = "pip install git+https://github.com/HumanSignal/label-studio-sdk.git"
```
The `label-studio-ml` the package is now detected. I can `pixi run` the ml backend.

It could be a more elegant solution, but it works in the docker container.

**Reply by matthewfeickert** [reply]
Glad that works well enough for you now. Though note that these dependencies aren't getting recorded in the `pypi-dependencies` table in `pixi.toml` or in the `pixi.lock`, and this makes it harder to take advantage of the [nice patterns for deployment of `pixi` environments to Linux containers](https://pixi.sh/latest/advanced/production_deployment/).

**Reply by SheldonWBM** [reply]
I am investigating using Pixi for that reason. I want to clean-up my dockerfile and have all packages installed through Pixi. 
Due to dependency version conflicts, the packages are messy with a specific and fragile install order. I am hoping I can use the pixi.toml to at least decrease the mess. The lable-studio packages are not in the pixi.lock but they are installed in the environment (not globally). I want one package manager to rule them all.

**Reply by dfang** [reply]
branches and tags are not supported yet on latest version pixi (0.39.5)

```
pixi add --pypi "dydx-v3-python @ git+https://github.com/konichuvak/dydx-v3-python.git@web3"
Error:
  x Found invalid characters for git revision 'web3', branches and tags are not supported yet
```

**Reply by matthewfeickert** [reply]
@dfang Yes, that would be nice to have. My recommendation is to follow the advice of Issue https://github.com/prefix-dev/pixi/issues/1281 and add a :+1: to Issue https://github.com/prefix-dev/pixi/issues/1078 where this is discussed.

**Reply by dfang** [reply]
> @dfang Yes, that would be nice to have. My recommendation is to follow the advice of Issue https://github.com/prefix-dev/pixi/issues/1281 and add a :+1: Issue https://github.com/prefix-dev/pixi/issues/1078 where this is discussed.

sure

**Comment by Norlandz** [comment]
**_Outdated_**:

<strike>
Failed on 

`pixi add --pypi "itables@git+https://github.com/mwouts/itables.git@robust_loading_of_offline_itables"`
`pixi add --pypi "itables @ git+https://github.com/mwouts/itables.git@robust_loading_of_offline_itables"`
`pixi add "itables@git+https://github.com/mwouts/itables.git@robust_loading_of_offline_itables"`
`pixi add "itables @ git+https://github.com/mwouts/itables.git@robust_loading_of_offline_itables"`

```
D:\usp\uhpsj\study\build_llm_wsp\build_llm>pixi add --pypi "itables@git+https://github.com/mwouts/itables.git@robust_loading_of_offline_itables"

thread 'main2' panicked at src\lock_file\package_identifier.rs:157:17:
internal error: entered unreachable code: git requirement on conda package is not supported
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

thread 'main' panicked at src\main.rs:49:10:
Tokio executor failed, was there a panic?: Any { .. }
```
</strike>

---

**_Update_**:

Need to update pixi

```
D:\usp\uhpsj\study\build_llm_wsp\build_llm>pixi self-update
Pixi will be updated from 0.47.0 to 0.53.0
Pixi archive downloaded.
Pixi archive uncompressed.
Pixi has been updated to version 0.53.0.
```

**Reply by lucascolley** [reply]
possibly fixed by https://github.com/prefix-dev/pixi/pull/4416
